core/src/main/scala/com/softwaremill/kmq/redelivery/streams/CommitMarkerStream.scala

```diff
@@ -0,0 +1,69 @@
+package com.softwaremill.kmq.redelivery.streams
+
+import akka.Done
+import akka.actor.ActorSystem
+import akka.kafka.ConsumerMessage.CommittableMessage
+import akka.kafka.scaladsl.Consumer.DrainingControl
+import akka.kafka.scaladsl.{Committer, Consumer}
+import akka.kafka.{CommitterSettings, ConsumerSettings, Subscriptions}
+import akka.stream.scaladsl.{Flow, Keep, Sink}
+import com.softwaremill.kmq.redelivery.Offset
+import com.softwaremill.kmq.{EndMarker, MarkerKey, MarkerValue, StartMarker}
+
+import scala.concurrent.Future
+
+class CommitMarkerStream(markerConsumerSettings: ConsumerSettings[MarkerKey, MarkerValue],
+                         markersTopic: String, maxPartitions: Int)
+                        (implicit system: ActorSystem) {
+  
+  def commitMarkerSink(): Sink[CommittableMessage[MarkerKey, MarkerValue], Future[Done]] = {
+    val committerSettings = CommitterSettings(system)
+
+    Flow[CommittableMessage[MarkerKey, MarkerValue]]
+      .statefulMapConcat { () => // keep track of open markers
+        val markersByOffset = new PriorityQueueMap[MarkerKey, CommittableMessage[MarkerKey, MarkerValue]](valueOrdering = bySmallestOffsetAscending)
+        msg => {
+          msg.record.value match {
+            case _: StartMarker => markersByOffset.put(msg.record.key, msg)
+            case _: EndMarker => markersByOffset.remove(msg.record.key)
+            case _ => throw new IllegalArgumentException()
+          }
+          markersByOffset.headOption
+        }
+      }
+      .statefulMapConcat { () => // deduplicate - pass only markers with increasing offsets
+        val maxOffset = new Ref[Offset]()
+        msg =>
+          if (!maxOffset.getOption.exists(_ >= msg.record.offset)) {
+            maxOffset.update(msg.record.offset)
+            Some(msg)
+          }
+          else None
+      }
+      .statefulMapConcat { () => // for each new marker return previous one
+        val previousMsg = new Ref[CommittableMessage[MarkerKey, MarkerValue]]()
+        msg =>
+          val prev = previousMsg.getOption
+          previousMsg.update(msg)
+          prev
+      }
+      .map(_.committableOffset)
+      .via(Committer.flow(committerSettings))
```

Comment:
something to verify: `consumer.commitSync(offset)` commits the given offset, which means that after a restart, the app will start processing data from the offset. So in other words, all messages up to `offset-1` are committed. 

Does `Committer.flow` work differently, that is it commits the given `offset+1`?

Reply:
You can look at the test: https://github.com/softwaremill/kmq/blob/24-refactor-redelivery-with-streams/core/src/test/scala/com/softwaremill/kmq/redelivery/streams/CommitMarkerSinkIntegrationTest.scala
Scenario (lines 53-75):
- saving startMarkers for messages 1 to 10
- saving endMarkers for 1, 2, 3, 5 (skipping 4)
- assert that client from other client group receives all markers
- assert that client from `kmqConfig.getRedeliveryConsumerGroupId` group receives markers 4 to 10 (meaning that 1 to 3 is successfully commited)

Reply:
ok sounds good :) thanks